### ISSUE TITLE
equivalent to scale layer in Caffe? And BN not supported in caffe_converter

### ISSUE THREAD

**Issue by shenxiul** [dmlc/mxnet#1473]
Now mxnet doesn't support batch normalization layer in caffe_converter, but BN is more or less a standard technique now. Any plan on that? Especially, caffe's implementation of BN is a bit non standard because it doesn't allow learning on beta and gamma. As a workaround MSRA uses a scale layer after BN layer to learn gamma and beta in ResNet, see https://github.com/KaimingHe/deep-residual-networks. Do we have something equivalent to caffe's scale layers in ResNet in mxnet? Now ResNet's pertained model cannot be converted by caffe_converter because of these two issues. Do we have any plan to resolve it recently? If not, I wish to help but not sure how to deal with the scale layer...

**Comment by winstywang** [comment]
In mxnet, BN layer learns the gamma (i.e. the slope), so it is equivalent to BN + scale layer in caffe

**Comment by shenxiul** [comment]
Hmm I just realized this part... Thanks @winstywang ! I'm working on resolving it now. Can I take that in mxnet we don't have such scale layer? I know for this specific problem we don't need it, just curious.

**Comment by piiswrong** [comment]
its equivalent to a 1x1 conv layer without bias right?

**Comment by shenxiul** [comment]
@piiswrong , awesome idea! But it's 1x1 conv layer *with* bias, right, as in BN layer we have both beta(bias) and gamma(scale)? what's the "groups" you mentioned?

**Comment by superzrx** [comment]
@shenxiul each conv kernel should only handle one input channel

**Comment by tqchen** [comment]
@shenxiul  The scaling layer you mentioned can be easily added, if necessary. Because the backend used mshadow expression template, and what we need is basically a simplified version of BN.

**Comment by tqchen** [comment]
Alternatively, we can add a broadcast operator and simply make it as 
```
out  = in * broadcast(bias, shape=out.shape, axis = 3)
```

**Comment by shenxiul** [comment]
@tqchen I'm not sure if it is necessary at this point, because to convert ResNet pre-trained model we only need scaling and bias added in BN. But it will be a good solution for converting caffe model in general, otherwise we might end up with some dirty workarounds all the time. I just want to make sure one thing, our BN layer only have scale term but not bias term?

**Comment by tqchen** [comment]
MXNet's BN layer have both scale and bias term, with the scale term can be optionally disabled via ```fix_gamma``` parameter

**Comment by shenxiul** [comment]
@tqchen I see. So if we don't have scale layer, we basically have to parse prototxt to check if BN connects with a scale layer and then save the parameters in scale layer into BN layer? It sounds not quite decent and if BN doesn't connect with Scale layer we don't have the option to fix bias...

**Comment by tqchen** [comment]
I get what you mean, you are more than welcomed to add a BroadcastMul Layer to mxnet, that resolves the issue.

**Comment by shenxiul** [comment]
@tqchen Sure! I'll let you know the progress. When I'm fairly sure I can handle this (I am new to mxnet and now learning its structure), I'll create a PR.

**Comment by shenxiul** [comment]
@tqchen It seems that my conversion from resnet pre-trained caffemodel works (at least symbol is expected), although I didn't do the scale layer yet. But to fine-tune it, I may need to change the upmost layer (softmax) to my desired loss function, and initialize fc1000 with random (not from pre-trained model because dimension doesn't match), also I may want to change the learning rate for the upmost layer for  a faster convergence. I looked at issue #1004 and you mentioned "internal" method can be used to extract internal parameters, however can we change it directly rather than build from scratch as the only thing we need to change is the top most layer? Or even if we have to build from scratch, do we have API to extract and iterate symbols from the original symbol to avoid do it by hand? Ultimately, can we have a tutorial/example for transfer learning from a pre-trained model as it is really a common use case? Thanks!

**Comment by tqchen** [comment]
see last block of https://github.com/dmlc/mxnet/blob/master/example/notebooks/predict-with-pretrained-model.ipynb

**Comment by shenxiul** [comment]
@tqchen, oh yes I misunderstood code in that block last night... Thanks!
